syntax = "proto3";

message RegisterCommunicator {
  string name = 1;
  string pid = 2;
}

message TrafficGeneratorParameters{
    string distribution_rate = 1;
    repeated float parameters_rate = 2;
    string distribution_execution_time = 3;
    repeated float parameters_execution_time = 4;
    string distribution_size = 5;
    repeated float parameters_size = 6;
}

message MessageRequest {
    string message = 1;
}

message SimulationInfo {
    string name = 1;
    int32 simulation_length = 2;
    float tick_length = 3;
    string ipaddress = 4;
}

message Counter {
    string actor_name = 1;
    string metric = 2;
    float value = 3;
}

message Counters {
    repeated Counter counters = 1;
}

message CreateActor {
    string type = 1;
    string name = 2;
    repeated float parameters = 3;
}

message CreateActors {
    repeated CreateActor create_actors = 1;
}

message RemoveActor {
    string type = 1;
    string name = 2;
}

message RemoveActors {
    repeated RemoveActor remove_actors = 1;
}

message UpdateParameterActor {
    string type = 1;
    string name = 2;
    string parameter_name = 3;
    float value = 4;
}

message ToPythonMessage {
    int32 transfer_id = 1;
    int32 tick_offset = 2;
    string pid_sender = 3;
    oneof payload {
        MessageRequest request = 4;
        SimulationInfo info = 5;
        RegisterCommunicator register_communicator = 6;
        Counters counters = 7;
    }
}

message ToSimulationMessage {
    string pid_receiver = 1;
    oneof payload {
        TrafficGeneratorParameters traffic_generator_params = 2;
        CreateActors create_actors = 4;
        RemoveActors remove_actors = 5;
        UpdateParameterActor update_parameter_actor = 6;
    }
}

message ResponseSimulation {
    int32 transfer_id = 1;
    repeated ToSimulationMessage messages = 2;
}

